File: typescript/day1.ts
import * as fs from "fs";
import * as path from "path";

// Define the relative path to the input file
const inputFilePath = path.join(__dirname, "../inputs/dayOneInput.txt");

// Function to read and process the input file
const processDay1Input = (filePath: string): number => {
    try {
        // Read the file content
        const data = fs.readFileSync(filePath, "utf-8");

        // Split the file into rows (lines)
        const rows = data.trim().split("\n");

        // Initialize variables
        let grandTotal = 0;
        const list1: number[] = [];
        const list2: number[] = [];

        // Process each row
        for (const row of rows) {
            // Split the row into two halves
            const middle = Math.round(row.length / 2);
            const firstHalf = parseInt(row.slice(0, middle), 10);
            const secondHalf = parseInt(row.slice(middle), 10);

            // Add to the respective lists
            list1.push(secondHalf);
            list2.push(firstHalf);
        }

        // Sort the lists
        list1.sort((a, b) => a - b);
        list2.sort((a, b) => a - b);

        // Calculate the grand total
        for (let i = 0; i < list1.length; i++) {
            const sumTotal = list1[i] - list2[i];
            grandTotal += Math.abs(sumTotal);
        }

        return grandTotal;
    } catch (error) {
        console.error("Error processing input file:", error);
        return 0;
    }
};

// Main execution
(() => {
    console.log("Processing Day 1 input...");

    // Process the input file and calculate the result
    const result = processDay1Input(inputFilePath);

    // Output the result
    console.log("Answer for Day 1 Part 1 with your input file is " + result);
})();

File: go/day1.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"
)

// Helper function to calculate the absolute value of an integer
func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func main() {
	// Define the path to your input file
	inputFilePath := "../inputs/dayOneInput.txt"

	// Open the file
	file, err := os.Open(inputFilePath)
	if err != nil {
		fmt.Println("Error opening file:", err)
		return
	}
	defer file.Close()

	// Initialize slices for the two lists
	var leftList []int
	var rightList []int

	// Read the input file line by line
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		// Split each line into two numbers
		line := strings.TrimSpace(scanner.Text())
		parts := strings.Fields(line)
		if len(parts) != 2 {
			fmt.Println("Invalid input format: each line must contain exactly two numbers")
			return
		}

		// Convert the numbers from strings to integers
		leftNum, err1 := strconv.Atoi(parts[0])
		rightNum, err2 := strconv.Atoi(parts[1])
		if err1 != nil || err2 != nil {
			fmt.Println("Error converting input to integers:", err1, err2)
			return
		}

		// Append the numbers to their respective lists
		leftList = append(leftList, leftNum)
		rightList = append(rightList, rightNum)
	}

	// Check for errors during file reading
	if err := scanner.Err(); err != nil {
		fmt.Println("Error reading file:", err)
		return
	}

	// Sort both lists
	sort.Ints(leftList)
	sort.Ints(rightList)

	// Calculate the total distance
	totalDistance := 0
	for i := 0; i < len(leftList); i++ {
		totalDistance += abs(leftList[i] - rightList[i])
	}

	// Print the result
	fmt.Printf("The total distance between the two lists is: %d\n", totalDistance)
}
